Crash in test repeat write 2 teardown

https://raw.githubusercontent.com/ThaliTester/TestResults/9937456522c40ea_Refactor_incoming_and_outgoing_socket_thread_tests_evabishchevich/build.md

find 
uncaught exception, error: 'AssertionError: entry exists

--- a/test/www/jxcore/bv_tests/testThaliNotificationAction.js
+++ b/test/www/jxcore/bv_tests/testThaliNotificationAction.js
@@ -115,21 +115,25 @@ var test = tape({
   setup: function (t) {
     sandbox = sinon.sandbox.create();
     globals = new GlobalVariables();
-    globals.init().then(function () {
-      t.end();
-    }).catch(function (failure) {
-      t.fail('Test setting up failed:' + failure);
-      t.end();
-    });
+    globals.init()
+      .then(function () {
+        t.end();
+      })
+      .catch(function (failure) {
+        t.fail('Test setting up failed:' + failure);
+        t.end();
+      });
   },
   teardown: function (t) {
     sandbox.restore();
-    globals.kill().then(function () {
-      t.end();
-    }).catch(function (failure) {
-      t.fail('Server cleaning failed:' + failure);
-      t.end();
-    });
+    globals.kill()
+      .then(function () {
+        t.end();
+      })
+      .catch(function (failure) {
+        t.fail('Server cleaning failed:' + failure);
+        t.end();
+      });
   }
 });
 
@@ -150,7 +154,8 @@ test('Test BEACONS_RETRIEVED_AND_PARSED locally', function (t) {
     addressBookCallback
   );
 
-  act.eventEmitter.on(NotificationAction.Events.Resolved,
+  act.eventEmitter.once(
+    NotificationAction.Events.Resolved,
     function (action, res, beaconDetails) {
       t.equals(
         action.getPeerIdentifier(),
@@ -173,16 +178,18 @@ test('Test BEACONS_RETRIEVED_AND_PARSED locally', function (t) {
           globals.preambleAndBeacons)) === 0, 'good preAmble');
       t.ok(globals.sourcePublicKeyHash.compare(
           beaconDetails.unencryptedKeyId) === 0, 'public keys match!');
-    });
+    }
+  );
 
-  act.start(globals.actionAgent).then(function (res) {
-    t.equals(res, null, 'must return null after successful call');
-    t.equals(act.getActionState(), ThaliPeerAction.actionState.KILLED,
-      'Once start returns the action should be in KILLED state');
-  })
-  .catch(function (failure) {
-    t.fail(failure.stack);
-  });
+  act.start(globals.actionAgent)
+    .then(function (res) {
+      t.equals(res, null, 'must return null after successful call');
+      t.equals(act.getActionState(), ThaliPeerAction.actionState.KILLED,
+        'Once start returns the action should be in KILLED state');
+    })
+    .catch(function (failure) {
+      t.fail(failure.stack);
+    });
 });
 
 test('Test HTTP_BAD_RESPONSE locally', function (t) {
@@ -190,7 +197,6 @@ test('Test HTTP_BAD_RESPONSE locally', function (t) {
 
   httpTester.runServer(globals.expressRouter,
     thaliConfig.NOTIFICATION_BEACON_PATH, 503, 'hello', 1);
-
 
   sandbox.stub(ThaliMobile, 'getPeerHostInfo')
     .withArgs(globals.testPeer.peerIdentifier, globals.testPeer.connectionType)
@@ -203,18 +209,23 @@ test('Test HTTP_BAD_RESPONSE locally', function (t) {
     TCP_NATIVE
   );
 
-  act.eventEmitter.on(NotificationAction.Events.Resolved, function (_, res) {
-    t.equals(
-      res, NotificationAction.ActionResolution.HTTP_BAD_RESPONSE,
-      'Response should be HTTP_BAD_RESPONSE'
-    );
-  });
+  act.eventEmitter.once(
+    NotificationAction.Events.Resolved,
+    function (_, res) {
+      t.equals(
+        res, NotificationAction.ActionResolution.HTTP_BAD_RESPONSE,
+        'Response should be HTTP_BAD_RESPONSE'
+      );
+    }
+  );
 
-  act.start(globals.actionAgent).then( function (res) {
-    t.equals(res, null, 'must return null after successful call');
-  }).catch(function (err) {
-    t.fail('Test failed:' + err.message);
-  });
+  act.start(globals.actionAgent)
+    .then(function (res) {
+      t.equals(res, null, 'must return null after successful call');
+    })
+    .catch(function (err) {
+      t.fail('Test failed:' + err.message);
+    });
 });
 
 test('Test NETWORK_PROBLEM locally', function (t) {
@@ -234,22 +245,27 @@ test('Test NETWORK_PROBLEM locally', function (t) {
     TCP_NATIVE
   );
 
-  act.eventEmitter.on(NotificationAction.Events.Resolved, function (_, res) {
-    t.equals(
-      res, NotificationAction.ActionResolution.NETWORK_PROBLEM,
-      'Response should be NETWORK_PROBLEM'
-    );
-  });
+  act.eventEmitter.once(
+    NotificationAction.Events.Resolved,
+    function (_, res) {
+      t.equals(
+        res, NotificationAction.ActionResolution.NETWORK_PROBLEM,
+        'Response should be NETWORK_PROBLEM'
+      );
+    }
+  );
 
-  act.start(globals.actionAgent).then(function () {
-    t.fail('This call should cause reject.');
-  }).catch(function (err) {
-    testUtils.restoreUnresolvableDomains();
-    t.equals(
-      err.message, 'Could not establish TCP connection',
-      'reject reason should be: Could not establish TCP connection'
-    );
-  });
+  act.start(globals.actionAgent)
+    .then(function () {
+      t.fail('This call should cause reject.');
+    })
+    .catch(function (err) {
+      testUtils.restoreUnresolvableDomains();
+      t.equals(
+        err.message, 'Could not establish TCP connection',
+        'reject reason should be: Could not establish TCP connection'
+      );
+    });
 });
 
 test('Action fails when getPeerHostInfo fails', function (t) {
@@ -272,18 +288,22 @@ test('Action fails when getPeerHostInfo fails', function (t) {
     TCP_NATIVE
   );
 
-  act.eventEmitter.on(NotificationAction.Events.Resolved, function (_, res) {
-    t.equals(
-      res, NotificationAction.ActionResolution.BAD_PEER,
-      'Resolution should be BAD_PEER'
-    );
-  });
+  act.eventEmitter.once(
+    NotificationAction.Events.Resolved,
+    function (_, res) {
+      t.equals(
+        res, NotificationAction.ActionResolution.BAD_PEER,
+        'Resolution should be BAD_PEER'
+      );
+    }
+  );
 
-  act.start(globals.actionAgent).then( function () {
-    t.fail('This call should cause reject.');
-  }).catch(function (err) {
-    t.equals(err.message, errorMessage, 'correct error message');
-  });
+  act.start(globals.actionAgent)
+    .then( function () {
+      t.fail('This call should cause reject.');
+    }).catch(function (err) {
+      t.equals(err.message, errorMessage, 'correct error message');
+    });
 });
 
 test('Call the start two times', function (t) {
@@ -304,31 +324,35 @@ test('Call the start two times', function (t) {
     TCP_NATIVE
   );
 
-  act.eventEmitter.on(NotificationAction.Events.Resolved, function (_, res) {
-    t.equals(
-      res,
-      NotificationAction.ActionResolution.BEACONS_RETRIEVED_AND_PARSED,
-      'Response should be BEACONS_RETRIEVED_AND_PARSED'
-    );
-  });
+  act.eventEmitter.once(
+    NotificationAction.Events.Resolved,
+    function (_, res) {
+      t.equals(
+        res,
+        NotificationAction.ActionResolution.BEACONS_RETRIEVED_AND_PARSED,
+        'Response should be BEACONS_RETRIEVED_AND_PARSED'
+      );
+    }
+  );
 
-
-  act.start(globals.actionAgent).then( function (res) {
+  act.start(globals.actionAgent)
+    .then( function (res) {
       t.equals(res, null, 'must return null after successful call.');
     })
     .catch(function (failure) {
       t.fail('Test failed:' + failure);
     });
 
-  act.start(globals.actionAgent).then( function () {
+  act.start(globals.actionAgent)
+    .then(function () {
       t.fail('Second start should not be successful.');
-  }).catch( function (err) {
-    t.equals(err.message, ThaliPeerAction.DOUBLE_START, 'Call start once');
-  });
+    })
+    .catch(function (err) {
+      t.equals(err.message, ThaliPeerAction.DOUBLE_START, 'Call start once');
+    });
 });
 
 test('Call the kill before calling the start', function (t) {
-
   t.plan(2);
 
   sandbox.stub(ThaliMobile, 'getPeerHostInfo')
@@ -342,22 +366,25 @@ test('Call the kill before calling the start', function (t) {
     TCP_NATIVE
   );
 
-  act.eventEmitter.on(NotificationAction.Events.Resolved, function (_, res) {
-    t.equals(
-      res, NotificationAction.ActionResolution.KILLED,
-      'Should be Killed'
-    );
-  });
+  act.eventEmitter.once(
+    NotificationAction.Events.Resolved,
+    function (_, res) {
+      t.equals(
+        res, NotificationAction.ActionResolution.KILLED,
+        'Should be Killed'
+      );
+    }
+  );
   act.kill();
 
-  act.start(globals.actionAgent).catch( function (err) {
-    t.equals(err.message, ThaliPeerAction.START_AFTER_KILLED,
-      'Start after killed');
-  });
+  act.start(globals.actionAgent)
+    .catch( function (err) {
+      t.equals(err.message, ThaliPeerAction.START_AFTER_KILLED,
+        'Start after killed');
+    });
 });
 
 test('Call the kill immediately after the start', function (t) {
-
   t.plan(2);
 
   // Sets 2000 milliseconds delay for request handling.
@@ -377,27 +404,29 @@ test('Call the kill immediately after the start', function (t) {
     TCP_NATIVE
   );
 
-  act.eventEmitter.on(NotificationAction.Events.Resolved, function (_, res) {
-    t.equals(
-      res,
-      NotificationAction.ActionResolution.KILLED,
-      'Should be KILLED'
-    );
-  });
+  act.eventEmitter.once(
+    NotificationAction.Events.Resolved,
+    function (_, res) {
+      t.equals(
+        res,
+        NotificationAction.ActionResolution.KILLED,
+        'Should be KILLED'
+      );
+    }
+  );
 
-  act.start(globals.actionAgent).then(function (res) {
-    t.equals(res, null, 'must return null after successful kill');
-  })
-  .catch(function (failure) {
-    t.fail('Test failed:' + failure);
-  });
+  act.start(globals.actionAgent)
+    .then(function (res) {
+      t.equals(res, null, 'must return null after successful kill');
+    })
+    .catch(function (failure) {
+      t.fail('Test failed:' + failure);
+    });
 
   act.kill();
-
 });
 
 test('Call the kill while waiting a response from the server', function (t) {
-
   t.plan(2);
 
   // Sets 10000 milliseconds delay for request handling.
@@ -418,22 +447,26 @@ test('Call the kill while waiting a response from the server', function (t) {
     TCP_NATIVE
   );
 
-  act.eventEmitter.on(NotificationAction.Events.Resolved, function (_, res) {
-    t.equals(
-      res, NotificationAction.ActionResolution.KILLED,
-      'Should be KILLED'
-    );
-  });
+  act.eventEmitter.once(
+    NotificationAction.Events.Resolved,
+    function (_, res) {
+      t.equals(
+        res, NotificationAction.ActionResolution.KILLED,
+        'Should be KILLED'
+      );
+    }
+  );
 
-
-  act.start(globals.actionAgent).then( function (res) {
-    t.equals(
-      res,
-      null,
-      'must return null after successful kill');
-  }).catch(function (err) {
-    t.fail('Test failed:' + err);
-  });
+  act.start(globals.actionAgent)
+    .then(function (res) {
+      t.equals(
+        res,
+        null,
+        'must return null after successful kill');
+    })
+    .catch(function (err) {
+      t.fail('Test failed:' + err);
+    });
 
   // This kills the action after 2 seconds. This should give enough time to
   // establish a HTTP connection in slow devices but since the server waits
@@ -443,11 +476,9 @@ test('Call the kill while waiting a response from the server', function (t) {
   setTimeout( function () {
     act.kill();
   }, 2000);
-
 });
 
 test('Test to exceed the max content size locally', function (t) {
-
   t.plan(2);
 
   var buffer = new Buffer(1024);
@@ -471,24 +502,28 @@ test('Test to exceed the max content size locally', function (t) {
     TCP_NATIVE
   );
 
-  act.eventEmitter.on(NotificationAction.Events.Resolved, function (_, res) {
-    t.equals(
-      res, NotificationAction.ActionResolution.HTTP_BAD_RESPONSE,
-      'HTTP_BAD_RESPONSE should be response when content size is exceeded'
-    );
-  });
+  act.eventEmitter.once(
+    NotificationAction.Events.Resolved,
+    function (_, res) {
+      t.equals(
+        res, NotificationAction.ActionResolution.HTTP_BAD_RESPONSE,
+        'HTTP_BAD_RESPONSE should be response when content size is exceeded'
+      );
+    }
+  );
 
-  act.start(globals.actionAgent).then( function (res) {
-    t.equals(res, null, 'must return null after successful call');
-  }).catch(function (failure) {
-    t.fail('Test failed:' + failure);
-  });
+  act.start(globals.actionAgent)
+    .then(function (res) {
+      t.equals(res, null, 'must return null after successful call');
+    })
+    .catch(function (failure) {
+      t.fail('Test failed:' + failure);
+    });
 });
 
 test('Close the server socket while the client is waiting a response ' +
   'from the server. Local test.',
   function (t) {
-
     t.plan(2);
 
     // Sets 10000 milliseconds delay for request handling.
@@ -512,22 +547,27 @@ test('Close the server socket while the client is waiting a response ' +
       TCP_NATIVE
     );
 
-    act.eventEmitter.on(NotificationAction.Events.Resolved, function (_, res) {
-      t.equals(
-        res, NotificationAction.ActionResolution.NETWORK_PROBLEM,
-        'Should be NETWORK_PROBLEM caused closing server socket'
-      );
-    });
+    act.eventEmitter.once(
+      NotificationAction.Events.Resolved,
+      function (_, res) {
+        t.equals(
+          res, NotificationAction.ActionResolution.NETWORK_PROBLEM,
+          'Should be NETWORK_PROBLEM caused closing server socket'
+        );
+      }
+    );
 
-    act.start(globals.actionAgent).then( function () {
-      t.fail('Test should return failure: Could not establish TCP connection');
-    }).catch(function (err) {
-      t.equals(
-        err.message,
-        'Could not establish TCP connection',
-        'Should be Could not establish TCP connection'
-      );
-    });
+    act.start(globals.actionAgent)
+      .then(function () {
+        t.fail('Test should return failure: Could not establish TCP connection');
+      })
+      .catch(function (err) {
+        t.equals(
+          err.message,
+          'Could not establish TCP connection',
+          'Should be Could not establish TCP connection'
+        );
+      });
 
     // This kills the server socket after 2 seconds. This should give enough
     // time to establish a HTTPS connection in slow devices but since the server
@@ -544,7 +584,6 @@ test('Close the server socket while the client is waiting a response ' +
 test('Close the client socket while the client is waiting a response ' +
   'from the server. Local test.',
   function (t) {
-
     t.plan(2);
 
     // Sets 10000 milliseconds delay for request handling.
@@ -567,21 +606,26 @@ test('Close the client socket while the client is waiting a response ' +
       TCP_NATIVE
     );
 
-    act.eventEmitter.on(NotificationAction.Events.Resolved, function (_, res) {
-      t.equals(
-        res, NotificationAction.ActionResolution.NETWORK_PROBLEM,
-        'Should be NETWORK_PROBLEM caused closing client socket'
-      );
-    });
+    act.eventEmitter.once(
+      NotificationAction.Events.Resolved,
+      function (_, res) {
+        t.equals(
+          res, NotificationAction.ActionResolution.NETWORK_PROBLEM,
+          'Should be NETWORK_PROBLEM caused closing client socket'
+        );
+      }
+    );
 
-    act.start(globals.actionAgent).then( function () {
-      t.fail('Test should return failure: Could not establish TCP connection');
-    }).catch(function (err) {
-      t.equals(
-        err.message,
-        'Could not establish TCP connection',
-        'Should be Could not establish TCP connection');
-    });
+    act.start(globals.actionAgent)
+      .then(function () {
+        t.fail('Test should return failure: Could not establish TCP connection');
+      })
+      .catch(function (err) {
+        t.equals(
+          err.message,
+          'Could not establish TCP connection',
+          'Should be Could not establish TCP connection');
+      });
 
     // This kills the client socket after 2 seconds. This should give enough
     // time to establish a HTTP connection in slow devices but since the server

--- a/test/www/jxcore/bv_tests/testThaliNotificationClient.js
+++ b/test/www/jxcore/bv_tests/testThaliNotificationClient.js
@@ -294,12 +294,6 @@ test('Received beacons with no values for us', function (t) {
     }
   );
 
-  var finalizeTest = function (err) {
-    getPeerHostInfoStub.restore();
-    enqueueStub.restore();
-    t.end(err);
-  };
-
   httpTester.runServer(globals.expressRouter,
     thaliConfig.NOTIFICATION_BEACON_PATH,
     200, globals.preambleAndBeacons, 1);
@@ -308,10 +302,24 @@ test('Received beacons with no values for us', function (t) {
     crypto.createECDH(thaliConfig.BEACON_CURVE).generateKeys();
   notificationClient.start([bogusPublicKey]);
 
-  notificationClient.on(notificationClient.Events.PeerAdvertisesDataForUs,
-    function () {
-      t.fail('We should not have gotten an event!');
-    });
+  notificationClient.on(
+    notificationClient.Events.PeerAdvertisesDataForUs,
+    notificationHandler
+  );
+  function notificationHandler() {
+    t.fail('We should not have gotten an event!');
+  }
+
+  var finalizeTest = function (err) {
+    getPeerHostInfoStub.restore();
+    enqueueStub.restore();
+
+    notificationClient.removeListener(
+      notificationClient.Events.PeerAdvertisesDataForUs,
+      notificationHandler
+    );
+    t.end(err);
+  };
 
   // New peer with TCP connection
   notificationClient._peerAvailabilityChanged(globals.TCPEvent);
@@ -338,6 +346,11 @@ test('Notification action killed with a superseded', function (t) {
         t.equal(entry.peerState, PeerDictionary.peerState.RESOLVED,
           'Peer state should be RESOLVED');
         peerPoolInterface.enqueue.restore();
+
+        notificationClient.removeListener(
+          notificationClient.Events.PeerAdvertisesDataForUs,
+          notificationHandler
+        );
         t.end();
       });
     });
@@ -353,10 +366,11 @@ test('Notification action killed with a superseded', function (t) {
 
   notificationClient.on(
     notificationClient.Events.PeerAdvertisesDataForUs,
-    function () {
-      t.fail('We should not have gotten a PeerAdvertisesDataForUs event!');
-    }
+    notificationHandler
   );
+  function notificationHandler() {
+    t.fail('We should not have gotten a PeerAdvertisesDataForUs event!');
+  }
 
   // New peer with TCP connection
   notificationClient._peerAvailabilityChanged(globals.TCPEvent);
@@ -379,13 +393,12 @@ test('Resolves an action locally', function (t) {
     function (action) {
       var keepAliveAgent = httpTester.getTestAgent(
         thaliConfig.BEACON_PSK_IDENTITY, thaliConfig.BEACON_KEY);
-      action.start(keepAliveAgent).then( function () {
-      }).catch( function ( ) {
-        t.fail('This action should not fail!');
-      });
+      action.start(keepAliveAgent)
+        .catch(function () {
+          t.fail('This action should not fail!');
+        });
     }
   );
-
 
 
   httpTester.runServer(globals.expressRouter,
@@ -398,7 +411,8 @@ test('Resolves an action locally', function (t) {
 
   notificationClient.start([globals.sourcePublicKey]);
 
-  notificationClient.on(notificationClient.Events.PeerAdvertisesDataForUs,
+  notificationClient.once(
+    notificationClient.Events.PeerAdvertisesDataForUs,
     function (res) {
       t.equals(
         res.hostAddress,
@@ -425,7 +439,8 @@ test('Resolves an action locally', function (t) {
       enqueueStub.restore();
       getPeerHostInfoStub.restore();
       t.end();
-    });
+    }
+  );
 
   // New peer with TCP connection
   notificationClient._peerAvailabilityChanged(globals.TCPEvent);
@@ -453,8 +468,9 @@ test('Resolves an action locally using ThaliPeerPoolDefault', function (t) {
 
   notificationClient.start([globals.sourcePublicKey]);
 
-  notificationClient.on(notificationClient.Events.PeerAdvertisesDataForUs,
-    function ( res) {
+  notificationClient.once(
+    notificationClient.Events.PeerAdvertisesDataForUs,
+    function (res) {
       t.equals(
         res.hostAddress,
         globals.TCPPeerHostInfo.hostAddress,
@@ -478,11 +494,11 @@ test('Resolves an action locally using ThaliPeerPoolDefault', function (t) {
 
       getPeerHostInfoStub.restore();
       t.end();
-    });
+    }
+  );
 
   // New peer with TCP connection
   notificationClient._peerAvailabilityChanged(globals.TCPEvent);
-
 });
 
 test('Action fails because of a bad hostname.', function (t) {
@@ -540,7 +556,7 @@ test('Action fails because of a bad hostname.', function (t) {
     function (action) {
       requestCount++;
       var keepAliveAgent = new http.Agent();
-      action.eventEmitter.on(
+      action.eventEmitter.once(
         ThaliNotificationAction.Events.Resolved,
         testResolutionEvent
       );
@@ -595,10 +611,10 @@ test('Action fails because of a bad hostname.', function (t) {
 
   // New peer with TCP connection
   notificationClient._peerAvailabilityChanged(TCPEvent);
-
 });
 
 test('hostaddress is removed when the action is running. ', function (t) {
+  var isTestFinished = false;
   // Scenario:
   // 1. Event: connectionType is TCP_NATIVE, peer is available
   // 2. Start to resolve the action
@@ -611,12 +627,16 @@ test('hostaddress is removed when the action is running. ', function (t) {
   // Simulates how peer pool runs actions
   var enqueue = function (action) {
     var keepAliveAgent = new http.Agent();
-    action.start(keepAliveAgent).then( function () {
+    action.start(keepAliveAgent)
+      .catch(function (err) {
+        if (isTestFinished) {
+          return;
+        }
+        isTestFinished = true;
 
-    }).catch( function (err) {
-      t.fail('This action should not fail');
-      t.end(err);
-    });
+        t.fail('This action should not fail');
+        t.end(err);
+      });
   };
 
   var enqueueStub = sinon.stub(globals.peerPoolInterface, 'enqueue', enqueue);
@@ -632,12 +652,22 @@ test('hostaddress is removed when the action is running. ', function (t) {
 
   notificationClient.start([globals.sourcePublicKey]);
 
-  notificationClient.on(notificationClient.Events.PeerAdvertisesDataForUs,
-    function () {
-      t.fail('This should never happen when action is getting killed' +
-        'because of the hostname is removed');
-      t.end();
-    });
+  notificationClient.once(
+    notificationClient.Events.PeerAdvertisesDataForUs,
+    notificationHandler
+  );
+  function notificationHandler() {
+    if (isTestFinished) {
+      return;
+    }
+    isTestFinished = true;
+
+    t.fail(
+      'This should never happen when action is getting killed ' +
+      'because of the hostname is removed'
+    );
+    t.end();
+  }
 
   // New peer with TCP connection
   notificationClient._peerAvailabilityChanged(globals.TCPEvent);
@@ -647,6 +677,11 @@ test('hostaddress is removed when the action is running. ', function (t) {
   // 10 seconds before it answers we have time to update the entry.
 
   setTimeout( function () {
+    if (isTestFinished) {
+      return;
+    }
+    isTestFinished = true;
+
     globals.TCPEvent.peerAvailable = false;
     notificationClient._peerAvailabilityChanged(globals.TCPEvent);
     t.equal(notificationClient.peerDictionary.size(), 0,
@@ -654,6 +689,7 @@ test('hostaddress is removed when the action is running. ', function (t) {
     notificationClient.stop();
     enqueueStub.restore();
     getPeerHostInfoStub.restore();
+
     t.end();
   }, 2000);
 });
@@ -712,56 +748,59 @@ test('notificationClient does not retry action with BAD_PEER resolution',
   }
 );
 
-test('notification client correctly handles peer availability changes ' +
-'of the same peer', function (t) {
-  testUtils.makeDomainUnresolvable('unresolvable_hostname');
-  globals.TCPPeerHostInfo.hostAddress = 'unresolvable_hostname';
+test(
+  'notification client correctly handles peer availability changes ' +
+  'of the same peer',
+  function (t) {
+    testUtils.makeDomainUnresolvable('unresolvable_hostname');
+    globals.TCPPeerHostInfo.hostAddress = 'unresolvable_hostname';
 
-  var notificationClient =
-    new ThaliNotificationClient(globals.peerPoolInterface,
-      globals.targetDeviceKeyExchangeObjects[0]);
+    var notificationClient =
+      new ThaliNotificationClient(globals.peerPoolInterface,
+        globals.targetDeviceKeyExchangeObjects[0]);
 
-  var callsCount = 0;
+    var callsCount = 0;
 
-  var enqueue = function (action) {
-    callsCount++;
+    var enqueue = function (action) {
+      callsCount++;
 
-    switch (callsCount) {
-      case 1:
-        var keepAliveAgent = new http.Agent({ keepAlive: true });
-        action.start(keepAliveAgent).then( function () {
-          t.fail('Should not succeed');
-        }).catch( function (err) {
-          // At this point action has fired Resolved event with network problem
-          // resolution and notificationClient scheduled retry for the failed
-          // action. We are going to fire new peerAvailabilityChanged event with
-          // the same peer id but different generation. Notification action
-          // should correctly handle this
-          t.pass('First action failed');
-          globals.TCPEvent.generation++;
-          notificationClient._peerAvailabilityChanged(globals.TCPEvent);
-        });
-        return;
-      case 2:
-        setTimeout(function () {
-          action.kill();
-          t.pass('second action killed');
-          enqueueStub.restore();
-          getPeerHostInfoStub.restore();
-          // notificationClient enqueues action and then synchronously updates
-          // peerDictionary with enqueued action. If we stop notificationClient
-          // inside enqueue stack it will fail.
-          notificationClient.stop();
-          testUtils.restoreUnresolvableDomains();
-          t.end();
-        }, ThaliNotificationClient.RETRY_TIMEOUTS[0] * 2);
-        return;
-    }
-  };
+      switch (callsCount) {
+        case 1:
+          var keepAliveAgent = new http.Agent({ keepAlive: true });
+          action.start(keepAliveAgent).then( function () {
+            t.fail('Should not succeed');
+          }).catch(function (err) {
+            // At this point action has fired Resolved event with network problem
+            // resolution and notificationClient scheduled retry for the failed
+            // action. We are going to fire new peerAvailabilityChanged event with
+            // the same peer id but different generation. Notification action
+            // should correctly handle this
+            t.pass('First action failed');
+            globals.TCPEvent.generation++;
+            notificationClient._peerAvailabilityChanged(globals.TCPEvent);
+          });
+          return;
+        case 2:
+          setTimeout(function () {
+            action.kill();
+            t.pass('second action killed');
+            enqueueStub.restore();
+            getPeerHostInfoStub.restore();
+            // notificationClient enqueues action and then synchronously updates
+            // peerDictionary with enqueued action. If we stop notificationClient
+            // inside enqueue stack it will fail.
+            notificationClient.stop();
+            testUtils.restoreUnresolvableDomains();
+            t.end();
+          }, ThaliNotificationClient.RETRY_TIMEOUTS[0] * 2);
+          return;
+      }
+    };
 
-  var enqueueStub = sinon.stub(globals.peerPoolInterface, 'enqueue', enqueue);
-  var getPeerHostInfoStub = stubGetPeerHostInfo();
+    var enqueueStub = sinon.stub(globals.peerPoolInterface, 'enqueue', enqueue);
+    var getPeerHostInfoStub = stubGetPeerHostInfo();
 
-  notificationClient.start([globals.sourcePublicKey]);
-  notificationClient._peerAvailabilityChanged(globals.TCPEvent);
-});
+    notificationClient.start([globals.sourcePublicKey]);
+    notificationClient._peerAvailabilityChanged(globals.TCPEvent);
+  }
+);

--- a/thali/NextGeneration/notification/thaliNotificationClient.js
+++ b/thali/NextGeneration/notification/thaliNotificationClient.js
@@ -180,7 +180,6 @@ ThaliNotificationClient.prototype.start =
  * @public
  */
 ThaliNotificationClient.prototype.stop = function () {
-
   if (this.peerDictionary) {
     ThaliMobile.emitter.removeListener('peerAvailabilityChanged',
       this._boundListener);

--- a/thali/NextGeneration/notification/thaliPeerDictionary.js
+++ b/thali/NextGeneration/notification/thaliPeerDictionary.js
@@ -128,27 +128,41 @@ PeerDictionary.MAXSIZE = 100;
  * Entry to be added.
   */
 PeerDictionary.prototype.addUpdateEntry =
-  function (peer, entry) {
-    var peerIdentifier = peer.peerIdentifier;
-    var generation = peer.generation;
+  function (newPeer, newEntry) {
+    var peerIdentifier = newPeer.peerIdentifier;
     assert(peerIdentifier, 'peer.peerIdentifier must be set');
+
+    var generation = newPeer.generation;
     assert(typeof generation === 'number', 'peer.generation must be a number');
 
-    if (!this._dictionary[peerIdentifier]) {
-      this._dictionary[peerIdentifier] = {};
+    var newPeerWillBeAdded = true;
+    var peerEntries = this._dictionary[peerIdentifier];
+    if (peerEntries) {
+      var oldPeer = peerEntries[generation];
+      if (oldPeer) {
+        // peer will be updated.
+        newPeerWillBeAdded = false;
+
+        if (oldPeer.entry != newEntry) {
+          this._removeEntry(oldPeer.entry);
+        }
+      }
+    }
+
+    if (newPeerWillBeAdded) {
+      this._removeOldestIfOverflow();
     }
 
     var peerEntries = this._dictionary[peerIdentifier];
-    if (!peerEntries[generation]) {
-      this._removeOldestIfOverflow();
-      peerEntries[generation] = {
-        peerIdentifier: peerIdentifier,
-        generation: generation,
-      };
+    if (!peerEntries) {
+      peerEntries = this._dictionary[peerIdentifier] = {};
     }
-
-    peerEntries[generation].entry = entry;
-    peerEntries[generation].entryNumber = this._entryCounter++;
+    peerEntries[generation] = {
+      peerIdentifier: peerIdentifier,
+      generation:     generation,
+      entry:          newEntry,
+      entryNumber:    this._entryCounter++
+    };
   };
 
 /**
@@ -173,11 +187,7 @@ PeerDictionary.prototype.remove = function (peer) {
   if (!entry) {
     return;
   }
-  entry.waitingTimeout && clearTimeout(entry.waitingTimeout);
-  entry.notificationAction &&
-    entry.notificationAction.eventEmitter.removeAllListeners(
-    ThaliNotificationAction.Events.Resolved);
-  entry.notificationAction && entry.notificationAction.kill();
+  this._removeEntry(entry);
 
   var peerEntries = this._dictionary[peer.peerIdentifier];
   delete peerEntries[peer.generation];
@@ -185,6 +195,18 @@ PeerDictionary.prototype.remove = function (peer) {
     delete this._dictionary[peer.peerIdentifier];
   }
 };
+
+PeerDictionary.prototype._removeEntry = function (entry) {
+  if (entry.waitingTimeout) {
+    clearTimeout(entry.waitingTimeout);
+  }
+  if (entry.notificationAction) {
+    entry.notificationAction.eventEmitter.removeAllListeners(
+      ThaliNotificationAction.Events.Resolved
+    );
+    entry.notificationAction.kill();
+  }
+}
 
 PeerDictionary.prototype.removeAllPeerEntries = function (peerIdentifier) {
   var peerEntries = this._dictionary[peerIdentifier];
@@ -323,7 +345,7 @@ PeerDictionary.prototype._removeOldestIfOverflow = function () {
     search(exports.peerState.CONTROLLED_BY_POOL);
 
   if (oldestPeer) {
-    self.remove(oldestPeer);
+    this.remove(oldestPeer);
   }
 };
 

